### PR TITLE
[TLI] Add mappings to SLEEF/ArmPL libcall variants taking linear args.

### PR DESCRIPTION
**Pull request by labrinea**
The mappings correspond to vectorized variants (fixed/scalable) for the math functions: modf, sincos, sincospi.

### CODE REVIEW

**Comment by mgabka** [comment]
could you also add tests for the replace-with-veclib-pass which is using those mapping to:
llvm/test/CodeGen/AArch64/replace-intrinsics-with-veclib-sleef-scalable.ll
llvm/test/CodeGen/AArch64/replace-intrinsics-with-veclib-sleef.ll
llvm/test/CodeGen/AArch64/replace-intrinsics-with-veclib-armpl.ll and

File: llvm/test/Transforms/LoopVectorize/AArch64/vector-libcall-linear-args.ll [hunk 1]
```diff
@@ -0,0 +1,275 @@
+; NOTE: Assertions have been autogenerated by utils/update_test_checks.py UTC_ARGS: --filter "call.*(modf|modff|sincos|sincosf|sincospi|sincospif)" --version 4
```
Comment (mgabka) [inline llvm/test/Transforms/LoopVectorize/AArch64/vector-libcall-linear-args.ll:1]: I was rather aiming to keep the tests in :
llvm/test/Transforms/LoopVectorize/AArch64/sleef-calls-aarch64.ll
llvm/test/Transforms/LoopVectorize/AArch64/armpl-calls.ll

so we have a single file per library, what do you think?

File: llvm/lib/Transforms/Vectorize/VPlanRecipes.cpp [hunk 2]
```diff
@@ -504,7 +504,8 @@ void VPWidenCallRecipe::execute(VPTransformState &State) {
   for (unsigned Part = 0; Part < State.UF; ++Part) {
     SmallVector<Type *, 2> TysForDecl;
     // Add return type if intrinsic is overloaded on it.
-    if (isVectorIntrinsicWithOverloadTypeAtArg(VectorIntrinsicID, -1)) {
+    if (isVectorIntrinsicWithOverloadTypeAtArg(VectorIntrinsicID, -1) &&
```
Comment (mgabka) [inline llvm/lib/Transforms/Vectorize/VPlanRecipes.cpp:507]: this is a LV bug, so I think it deserves a separate commit with target independent test, probably extension to the tests added in https://github.com/llvm/llvm-project/pull/73941 as all of the tests there were using functions  returning a value.

File: llvm/test/Transforms/LoopVectorize/AArch64/vector-libcall-linear-args.ll (hunk 1 above)
Comment (labrinea) [reply llvm/test/Transforms/LoopVectorize/AArch64/vector-libcall-linear-args.ll:1]: The mentioned test files have been refactored in #76146.

File: llvm/lib/Transforms/Vectorize/VPlanRecipes.cpp (hunk 2 above)
Comment (labrinea) [reply llvm/lib/Transforms/Vectorize/VPlanRecipes.cpp:507]: This fix has been moved to a separate PR as suggested: #76274

**Comment by labrinea** [comment]
> could you also add tests for the replace-with-veclib-pass which is using those mapping to: llvm/test/CodeGen/AArch64/replace-intrinsics-with-veclib-sleef-scalable.ll llvm/test/CodeGen/AArch64/replace-intrinsics-with-veclib-sleef.ll llvm/test/CodeGen/AArch64/replace-intrinsics-with-veclib-armpl.ll

I am not seeing corresponding llvm instrinsics for the veclib functions I've added. I don't think there's anything needed to be done in theses test files.

File: llvm/test/Transforms/LoopVectorize/AArch64/vector-libcall-linear-args.ll [hunk 3]
```diff
@@ -0,0 +1,275 @@
...
+  br label %for.body
+
+for.body:
+  %indvars.iv = phi i64 [ 0, %entry ], [ %indvars.iv.next, %for.body ]
+  %gepa = getelementptr double, ptr %a, i64 %indvars.iv
+  %num = load double, ptr %gepa, align 8
+  %gepb = getelementptr double, ptr %b, i64 %indvars.iv
+  %data = call double @modf(double %num, ptr %gepb)
+  %gepc = getelementptr inbounds double, ptr %c, i64 %indvars.iv
+  store double %data, ptr %gepc, align 8
```
Comment (mgabka) [inline llvm/test/Transforms/LoopVectorize/AArch64/vector-libcall-linear-args.ll:45]: this isn't needed

File: llvm/test/Transforms/LoopVectorize/AArch64/vector-libcall-linear-args.ll [hunk 4]
```diff
@@ -0,0 +1,275 @@
...
+; RUN: opt < %s -mattr=+sve -vector-library=sleefgnuabi -passes=inject-tli-mappings,loop-vectorize -S | FileCheck %s --check-prefix=SLEEF-SVE
+; RUN: opt < %s -mattr=+neon -vector-library=ArmPL -passes=inject-tli-mappings,loop-vectorize -S | FileCheck %s --check-prefix=ARMPL-NEON
+; RUN: opt < %s -mattr=+sve -vector-library=ArmPL -passes=inject-tli-mappings,loop-vectorize -S | FileCheck %s --check-prefix=ARMPL-SVE
+
+target triple = "aarch64-unknown-linux-gnu"
+
+define void @test_modf(ptr noalias %a, ptr noalias %b, ptr noalias %c, i64 %n) {
+; SLEEF-NEON-LABEL: define void @test_modf(
+; SLEEF-NEON-SAME: ptr noalias [[A:%.*]], ptr noalias [[B:%.*]], ptr noalias [[C:%.*]], i64 [[N:%.*]]) #[[ATTR0:[0-9]+]] {
+; SLEEF-NEON:    [[TMP8:%.*]] = call <2 x double> @_ZGVnN2vl8_modf(<2 x double> [[WIDE_LOAD:%.*]], <2 x ptr> [[TMP6:%.*]])
```
Comment (mgabka) [inline llvm/test/Transforms/LoopVectorize/AArch64/vector-libcall-linear-args.ll:13]: these tests are just showing that there is a compiler bug.

In this particular case you can see that it is passing <2x ptr> to the function instead of just ptr, the whole idea of linear pointers is that you are passing just a ptr to the function and the function will increment it inside according to the step.

I am not sure where is the bug, the tests added by @huntergr-arm  were using only integer types from what I can see https://github.com/llvm/llvm-project/pull/73941/files#diff-9e8f66d7297707a074a06db9f068f2bd0cff25d12e4fc07def062d224416aff4

Could you please investigate it?

File: llvm/test/Transforms/LoopVectorize/AArch64/vector-libcall-linear-args.ll (hunk 3 above)
Comment (labrinea) [reply llvm/test/Transforms/LoopVectorize/AArch64/vector-libcall-linear-args.ll]: Without a store we are not using the return value of the call. I see no harm here.

File: llvm/test/Transforms/LoopVectorize/AArch64/vector-libcall-linear-args.ll (hunk 4 above)
Comment (labrinea) [reply llvm/test/Transforms/LoopVectorize/AArch64/vector-libcall-linear-args.ll]: Fixed in https://github.com/llvm/llvm-project/pull/76753

File: llvm/test/Transforms/LoopVectorize/AArch64/vector-libcall-linear-args.ll (hunk 3 above)
Comment (mgabka) [reply llvm/test/Transforms/LoopVectorize/AArch64/vector-libcall-linear-args.ll]: sure, I just pointed it out as it could simplify the test.

**Review by mgabka — APPROVED** [review]

**Review by paschalis-mpeis — APPROVED** [review]
